crypto/stark-verifier-ethereum/contracts/stark_verifier.sol

```diff
@@ -0,0 +1,193 @@
+    // NOTE - This function also advances the coin by writing the pow_nonce to it
+    function check_proof_of_work(PublicCoin.Coin memory coin, bytes8 pow_nonce, uint8 pow_bits)
+        internal
+        pure
+        returns (bool)
+    {
+        bytes32 rand_from_channel = coin.read_bytes32();
+        bytes32 seed = keccak256(abi.encodePacked(hex'0123456789abcded', rand_from_channel, pow_bits));
+        bytes32 response = keccak256(abi.encodePacked(seed, pow_nonce));
+        uint8 leading_zeros = leading_zeros(response);
+        coin.write_bytes(abi.encodePacked(pow_nonce));
+        return leading_zeros >= pow_bits;
+    }
+
+    // Returns the number of leading zeros in a bytes32 data
+    function leading_zeros(bytes32 data) internal pure returns (uint8) {
+        // One set in the leading bit position
+        bytes32 leading_one = 0x8000000000000000000000000000000000000000000000000000000000000000;
+        uint8 result = 0;
+        while (leading_one & data == 0) {
+            leading_one >> 1;
+            result++;
+            // We don't want a revert if the zero is passed in
+            if (result == 255) {
+                return result;
+            }
+        }
+        return result;
+    }
+
+    // Reads from channel random and returns a list of random queries
+    function get_queries(PublicCoin.Coin memory coin, uint8 max_bit_length, uint8 num_queries)
+        internal
+        view
+        returns (uint64[] memory)
+    {
+        uint64[] memory queries = new uint64[](num_queries);
+        // This mask sets all digits to one below the bit length
+        uint64 bit_mask = (uint64(2)**max_bit_length) - 1;
+
+        // We derive four queries from each read
+        for (uint256 i = 0; i <= num_queries / 4; i++) {
+            bytes32 random = coin.read_bytes32();
+            for (uint256 j = 0; j < 4; j++) {
+                // For numbers of queries which are not diviable by four this prevents writing out of bounds.
+                if (4 * i + j < num_queries) {
+                    // Note - uint64(random) would take the last bytes in the random and this takes the first.
+                    queries[4 * i + j] = uint64(bytes8(random)) & bit_mask;
+                    // Shifts down so we can get the next set of random bytes
+                    random <<= 64;
+                }
+            }
+        }
+        sort(queries);
+        return queries;
+    }
+
+    // This function sorts the array
+    // Note - We use insertion sort, the array is expected to be small so this shouldn't
+    // cause problems.
+    function sort(uint64[] memory data) internal pure {
+        for (uint256 i = 0; i < data.length; i++) {
+            uint256 j = i;
+            while (j > 0 && data[j] < data[j - 1]) {
+                uint64 held = data[j];
+                data[j] = data[j - 1];
+                data[j - 1] = held;
```

Comment:
Solidity has single line swaps I believe. It would look like:

```solidity
(data[j], data[j - 1]) = (data[j], data[j - 1]);
```

Reply:
And apparently there's a slightly more efficient (not that this currently matters) version:

```
i ← 1
while i < length(A)
    x ← A[i]
    j ← i - 1
    while j >= 0 and A[j] > x
        A[j+1] ← A[j]
        j ← j - 1
    end while
    A[j+1] ← x[3]
    i ← i + 1
end while
```

Reply:
I played with implementing the slightly faster version, but it got annoying because of the assumption that j is a signed index and solidity doesn't like that. Decided to just leave the slightly worse one since the difference should very small.